### PR TITLE
[Inference]Adapt to baichuan2 13B

### PR DESCRIPTION
**Pull request by yuehuayingxueluo**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
pytest:
image

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/modeling/policy/nopadding_baichuan.py [hunk 1]
```diff
@@ -21,14 +24,13 @@ def module_policy(self):
         policy = super().module_policy()
 
         decoder_attribute_replacement = {
-            "lm_head.weight": Parameter(
-                nn.functional.normalize(self.model.lm_head.weight).transpose(0, 1), requires_grad=False
-            ),
+            "lm_head.weight": Parameter(nn.functional.normalize(self.model.lm_head.weight), requires_grad=False),
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/policy/nopadding_baichuan.py:27]: Wonder if the change aims at supporting TP?

File: colossalai/inference/modeling/policy/nopadding_baichuan.py [hunk 2]
```diff
@@ -42,17 +44,38 @@ def module_policy(self):
...
+            sub_module_replacement=[
+                SubModuleReplacementDescription(
+                    suffix="mlp",
+                    target_module=NopadBaichuanMLP,
+                ),
+                SubModuleReplacementDescription(
+                    suffix="self_attn",
+                    target_module=NopadBaichuanAttention,
+                ),
+            ]
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/policy/nopadding_baichuan.py:58]: Please extract the replacement description to reduce redundancy. For now the two layers use the same sub module replacement description.

File: colossalai/inference/modeling/policy/nopadding_baichuan.py [hunk 3]
```diff
@@ -42,17 +44,38 @@ def module_policy(self):
...
             description={"forward": llama_model_forward}, policy=policy, target_key="BaichuanModel"
         )
+
+        # used for Baichuan 7B
         self.append_or_create_method_replacement(
             description={"forward": llama_decoder_layer_forward}, policy=policy, target_key="DecoderLayer"
         )
+        # used for Baichuan 13B
+        self.append_or_create_method_replacement(
+            description={"forward": llama_decoder_layer_forward}, policy=policy, target_key="BaichuanLayer"
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/policy/nopadding_baichuan.py:74]: Same as above, though the number of lines keep unchanged.

File: tests/test_infer/test_models/test_baichuan.py [hunk 4]
```diff
@@ -39,7 +40,7 @@ def check_inference_engine(use_engine=False, prompt_template=None):
 
     if use_engine:
         inference_config = InferenceConfig(
-            max_output_len=output_len, prompt_template=prompt_template, dtype="fp32", use_cuda_kernel=True
+            max_output_len=output_len, prompt_template=prompt_template, use_cuda_kernel=False
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_models/test_baichuan.py:43]: Is it related with rms norm, which has been fixed in #5611 ?

File: tests/test_infer/test_ops/triton/kernel_utils.py [hunk 5]
```diff
@@ -64,10 +64,12 @@ def torch_attn_ref(
...
-        if attention_mask.size() != (bsz, 1, q_len, kv_len):
-            raise ValueError(
-                f"Attention mask should be of size {(bsz, 1, q_len, kv_len)}, but is {attention_mask.size()}"
-            )
+        if q_len == 1:
+            if len(attention_mask.size()) == 4:
+                attention_mask = attention_mask[:, :, -1:, :]
+            else:
+                attention_mask = attention_mask[:, -1:, :]
+
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_ops/triton/kernel_utils.py:72]: Wonder why adding this part? Attention masks with unusual forms should be transformed before the `torch_attn_ref`. If alibi has something to do with attention mask, put the logic outside this reference function.

File: tests/test_infer/test_ops/triton/test_context_attn_unpad.py [hunk 6]
```diff
@@ -19,8 +20,31 @@
 HEAD_DIM = 32
 
 
+def _fill_with_neg_inf(t):
+    return t.float().fill_(float("-inf")).type_as(t)
+
+
+# alibi mask calculation adapted from https://huggingface.co/baichuan-inc/Baichuan2-13B-Chat/blob/main/modeling_baichuan.py
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_ops/triton/test_context_attn_unpad.py:27]: Same as above, use ref transformers v4.36 if it has this function

File: tests/test_infer/test_ops/triton/test_decoding_attn.py [hunk 7]
```diff
@@ -130,14 +155,16 @@ def test_flash_decoding(
...
+        alibi_slopes=alibi_slopes,
         sm_scale=sm_scale,
         kv_group_num=kv_group_num,
         q_len=q_len,
     )  # [bsz * q_len, num_heads, head_dim]
 
     assert out_torch.shape == out_triton.shape
-    assert torch.allclose(out_torch, out_triton, atol=1e-3, rtol=1e-4)
+    # After the shape becomes larger, some data elements are too small, leading to excessively large relative errors.
+    numpy_allclose(out_torch, out_triton, atol=1e-3, rtol=100)
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_ops/triton/test_decoding_attn.py:166]: `rtol=100`, typo?

File: colossalai/kernel/triton/flash_decoding.py [hunk 8]
```diff
@@ -124,6 +124,129 @@ def _flash_decoding_fwd_kernel(
...
+@triton.jit
+def _alibi_flash_decoding_fwd_kernel(
+    Q,  # [batch_size * q_len, head_num, head_dim]
+    KCache,  # [num_blocks, num_kv_heads, block_size, head_dim]
+    VCache,  # [num_blocks, num_kv_heads, block_size, head_dim]
+    block_tables,  # [batch_size, max_blocks_per_sequence]
+    mid_o,  # [batch_size * q_len, head_num, kv_split_num, head_dim]
+    mid_o_lse,  # [batch_size * q_len, head_num, kv_split_num]
+    kv_seq_len,  # [batch_size]
+    q_len,
```
Comment (yuanheng-zhao) [inline colossalai/kernel/triton/flash_decoding.py:137]: Please add a NOTE or TODO to notify people that alibi flash decoding does not support q_len > 1 at this moment

File: colossalai/inference/modeling/models/nopadding_baichuan.py [hunk 9]
```diff
@@ -5,15 +5,61 @@
...
 
+def baichuan_rmsnorm_forward(
+    self,
+    hidden_states: torch.Tensor,
+    norm_output: torch.Tensor,
+    residual: torch.Tensor = None,
+    use_cuda_kernel: bool = True,
+):
+    # Used to address the issue of inconsistent epsilon variable names in baichuan2 7b and 13b.
+    if hasattr(self, "variance_epsilon"):
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_baichuan.py:42]: What I'm thinking is that if Baichuan releases a 70B model and they change the variable names again.

File: colossalai/inference/modeling/policy/nopadding_baichuan.py (hunk 1 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/policy/nopadding_baichuan.py:27]: YES

File: tests/test_infer/test_ops/triton/test_decoding_attn.py (hunk 7 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_ops/triton/test_decoding_attn.py:166]: After the shape becomes larger, the relative error of some shapes can reach up to 30.

File: colossalai/inference/flash_decoding_utils.py [hunk 10]
```diff
@@ -1,9 +1,27 @@
+import math
+
 import torch
 
 from colossalai.context.singleton_meta import SingletonMeta
 from colossalai.utils import get_current_device
 
 
+# alibi slopes calculation adapted from https://github.com/huggingface/transformers/blob/v4.39.3/src/transformers/models/bloom/modeling_bloom.py#L57
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/flash_decoding_utils.py:9]: There is no difference between v4.39 and v4.36.

File: tests/test_infer/test_ops/triton/kernel_utils.py (hunk 5 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_ops/triton/kernel_utils.py:72]: After adding alibi_mask, the shape of the last two dimensions of attention_mask will become (max_seq_len, max_seq_len).

File: tests/test_infer/test_ops/triton/test_context_attn_unpad.py (hunk 6 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_ops/triton/test_context_attn_unpad.py:27]: This function is adapted from Baichuan modeling.

File: colossalai/inference/flash_decoding_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/flash_decoding_utils.py:9]: Then revise the comment - use the url to v4.36, which is consistent with dependency version

File: colossalai/inference/modeling/models/nopadding_baichuan.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/modeling/models/nopadding_baichuan.py:42]: Hence, at this moment, only two possibilities exist (7B/13B). Please comply with Python logic and styling. Change to `elif` when there have already been more options.

[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/modeling/models/nopadding_baichuan.py:42]: Alternatively, you could comply with Python logic and styling by adding `else` to remind the user unexpected attribute found. **The idea here is to handle all the conditions** when we use if-else blocks or if-elif-else blocks, instead of allowing some condition fleeing away and cause unexpected error.

File: tests/test_infer/test_ops/triton/kernel_utils.py (hunk 5 above)
Comment (yuanheng-zhao) [reply tests/test_infer/test_ops/triton/kernel_utils.py:72]: Can you put the logic outside the function `torch_attn_ref`? If any specific transformations exist for a specific test, put the logic in that test.

File: tests/test_infer/test_ops/triton/test_decoding_attn.py (hunk 7 above)
Comment (yuanheng-zhao) [reply tests/test_infer/test_ops/triton/test_decoding_attn.py:166]: Revert to `rtol=1e-4`. You could split those conditions with other for now.

File: colossalai/inference/modeling/models/nopadding_baichuan.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_baichuan.py:42]: OK

File: tests/test_infer/test_ops/triton/kernel_utils.py (hunk 5 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_ops/triton/kernel_utils.py:72]: No, the original assert statement sets num_head to 1 because the masks for all head_dims in the head dimension are consistent and can be automatically expanded. However, in alibi flash attn, the mask for each head's head dim is not consistent, preventing automatic expansion. After adding the alibi mask, the masks for the head dims corresponding to all heads will be generated, eliminating the need for one-dimensional expansion.

File: colossalai/inference/core/request_handler.py [hunk 11]
```diff
@@ -111,7 +111,7 @@ class RequestHandler:
        dtype (torch.dtype): The data type for weights and activations.
     """
 
-    def __init__(self, inference_config: InferenceConfig, model_config: PretrainedConfig) -> None:
+    def __init__(self, inference_config: InferenceConfig, model_config: PretrainedConfig, alibi_attn: bool) -> None:
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/core/request_handler.py]: Done

File: colossalai/inference/modeling/models/nopadding_baichuan.py [hunk 12]
```diff
@@ -5,15 +5,61 @@
...
+        eps = self.variance_epsilon
+    elif hasattr(self, "epsilon"):
+        eps = self.epsilon
+
+    # TODO "Currently, CUDA RMS Norm does not support the hidden_size of 5120. Remove this code after CUDA RMS Norm is fixed."
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_baichuan.py]: Done

File: colossalai/inference/modeling/policy/nopadding_baichuan.py (hunk 2 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/policy/nopadding_baichuan.py]: Done

File: colossalai/inference/modeling/policy/nopadding_baichuan.py (hunk 3 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/policy/nopadding_baichuan.py]: Done

File: tests/test_infer/test_models/test_baichuan.py (hunk 4 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_models/test_baichuan.py]: Done

File: tests/test_infer/test_ops/triton/test_decoding_attn.py (hunk 7 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_ops/triton/test_decoding_attn.py]: Done

File: colossalai/kernel/triton/flash_decoding.py (hunk 8 above)
Comment (yuehuayingxueluo) [reply colossalai/kernel/triton/flash_decoding.py:137]: Done

File: tests/test_infer/test_ops/triton/kernel_utils.py (hunk 5 above)
Comment (yuanheng-zhao) [reply tests/test_infer/test_ops/triton/kernel_utils.py:72]: From your description and your code, I think you could prevent adding the if block in this function. Just put the transformation outside `torch_attn_ref`

Comment (yuehuayingxueluo) [reply tests/test_infer/test_ops/triton/kernel_utils.py:72]: However, the assert statement mentioned above may need to be deleted.

File: colossalai/inference/flash_decoding_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply colossalai/inference/flash_decoding_utils.py]: Done

File: tests/test_infer/test_ops/triton/kernel_utils.py (hunk 5 above)
Comment (yuanheng-zhao) [reply tests/test_infer/test_ops/triton/kernel_utils.py:72]: "I think you could **prevent adding the if block in this function**. Just put the transformation outside torch_attn_ref"
--------
Please at least read the whole sentence...

**Review by yuanheng-zhao — APPROVED** [review]
